azure/services/privatelinks/spec.go

```diff
@@ -0,0 +1,239 @@
+	Subnet string
+
+	// PrivateIPAddress is the optional static private IP address from the specified Subnet.
+	PrivateIPAddress string
+}
+
+// ResourceName returns the name of the NAT gateway.
+func (s *PrivateLinkSpec) ResourceName() string {
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the resource group.
+func (s *PrivateLinkSpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName is a no-op for NAT gateways.
+func (s *PrivateLinkSpec) OwnerResourceName() string {
+	return ""
+}
+
+// Parameters returns the parameters for the NAT gateway.
+func (s *PrivateLinkSpec) Parameters(ctx context.Context, existing interface{}) (params interface{}, err error) {
+	if len(s.NATIPConfiguration) == 0 {
+		return nil, errors.Errorf("At least one private link NAT IP configuration must be specified")
+	}
+	if len(s.LBFrontendIPConfigNames) == 0 {
+		return nil, errors.Errorf("At least one load balancer front end name must be specified")
+	}
+
+	// NAT IP configurations
+	var ipConfigurations []network.PrivateLinkServiceIPConfiguration
+	for i, natIPConfiguration := range s.NATIPConfiguration {
+		ipAllocationMethod := network.IPAllocationMethod(natIPConfiguration.AllocationMethod)
+		if ipAllocationMethod != network.Dynamic && ipAllocationMethod != network.Static {
+			return nil, errors.Errorf("%T is not a supported network.IPAllocationMethodStatic", natIPConfiguration.AllocationMethod)
+		}
+		var privateIpAddress *string
+		if ipAllocationMethod == network.Static {
+			if natIPConfiguration.PrivateIPAddress != "" {
+				privateIpAddress = pointer.String(natIPConfiguration.PrivateIPAddress)
+			} else {
+				return nil, errors.Errorf("Private link NAT IP configuration with static IP allocation must specify a private address")
+			}
+		}
+		ipConfiguration := network.PrivateLinkServiceIPConfiguration{
+			Name: pointer.String(fmt.Sprintf("%s-natipconfig-%d", natIPConfiguration.Subnet, i+1)),
+			PrivateLinkServiceIPConfigurationProperties: &network.PrivateLinkServiceIPConfigurationProperties{
+				Subnet: &network.Subnet{
+					ID: pointer.String(
+						fmt.Sprintf(
+							"/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s/subnets/%s",
+							s.SubscriptionID,
+							s.ResourceGroup,
+							s.VNet,
+							natIPConfiguration.Subnet)),
+				},
+				PrivateIPAllocationMethod: ipAllocationMethod,
+				PrivateIPAddress:          privateIpAddress,
+			},
+		}
+		ipConfigurations = append(ipConfigurations, ipConfiguration)
+		ipConfigurations[0].Primary = pointer.Bool(true)
+	}
+
+	// Load balancer front-end IP configurations
+	var frontendIPConfigurations []network.FrontendIPConfiguration
+	for _, frontendIPConfigName := range s.LBFrontendIPConfigNames {
+		frontendIPConfig := network.FrontendIPConfiguration{
+			ID: pointer.String(
+				fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/loadBalancers/%s/frontendIPConfigurations/%s",
```

Comment:
You should be able to use `FrontendIPConfigID()` in [this file](https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/b07d34e72a9ff577c77342b7eb5d6c610c71335c/azure/defaults.go#L246-L249) instead.